test/distributed/test_c10d_nccl.py

```diff
@@ -4228,6 +4251,140 @@ def test_nccl_errors_dump(self):
+        c10d.init_process_group(
+            "nccl",
+            world_size=self.world_size,
+            rank=self.rank,
+            store=store,
+            pg_options=opts,
+            device_id=device_id,
+        )
+        pg = c10d.distributed_c10d._get_default_group()
+        return pg
+
+    def opts(self, high_priority_stream=False):
+        opts = c10d.ProcessGroupNCCL.Options()
+        opts.is_high_priority_stream = high_priority_stream
+        return opts
+
+    def setUp(self):
+        super().setUp()
+        # TORCH_NCCL_BLOCKING_WAIT overrides TORCH_NCCL_ASYNC_ERROR_HANDLING hence tests
+        # that use TORCH_NCCL_BLOCKING_WAIT will test it as expected.
+        os.environ["TORCH_NCCL_ASYNC_ERROR_HANDLING"] = "1"
+        # self.num_gpus = torch.cuda.device_count()
+        self._spawn_processes()
+
+    def tearDown(self):
+        super().tearDown()
+        try:
+            os.remove(self.file_name)
+        except OSError:
+            pass
+
+    @property
+    def world_size(self):
+        return 8
+
+    @property
+    def rank_to_GPU(self):
+        # return rank to GPU map
+        return init_multigpu_helper(self.world_size, "nccl")
+
+    @requires_nccl_version((2, 18), "Need NCCL 2.18+ for ncclCommSplit")
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "NCCL test requires 2+ GPUs")
+    @skip_if_lt_x_gpu(8)
+    def test_comm_split_group_larger_scale(self):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        device = torch.device(f"cuda:{self.rank}")
+        pg = self._create_process_group_nccl(store, self.opts(), device_id=device)
+        backend = pg._get_backend(torch.device(device))
+
+        tensor = torch.full((1,), self.rank).cuda(device)
+        ng1 = c10d.split_group(pg, [[0, 1], [2, 3, 4, 5, 6, 7]])
+        backend1 = ng1._get_backend(torch.device(device))
+
+        # comm split happens eagerly since device_id is passed to init_process_group.
+        self.assertEqual(backend.comm_split_count(), 1)
+        # dist.broadcast take Source rank on global process group
+        if self.rank < 2:
+            dist.broadcast(tensor, 0, group=ng1)
+            self.assertEqual(tensor, torch.full((1,), 0))
+        else:
+            dist.broadcast(tensor, 2, group=ng1)
+            self.assertEqual(tensor, torch.full((1,), 2))
+
+        # test split with only one colored group, other ranks should be no color split.
+        ng2 = c10d.split_group(pg, [[5, 6, 7]], timeout=timedelta(seconds=50))
+        self.assertEqual(backend.comm_split_count(), 2)
+
+        if self.rank >= 5:
+            tensor2 = torch.full((1,), self.rank).cuda(device)
+            dist.broadcast(tensor2, 7, group=ng2)
+            self.assertEqual(tensor2, torch.full((1,), 7))
+        else:
+            self.assertEqual(ng2, c10d.GroupMember.NON_GROUP_MEMBER)
+        # give enough time for the broadcast to finish before destroying all pgs.
+        time.sleep(2)
```

Comment:
can't you just wait the broadcast and synchronize cuda? then avoid sleep?

Reply:
weirdly, broadcast() should be sync call by default, which means work.wait is called. But obviously it does not work as expected, this is something we need to follow up

Reply:
No, it works as expected but you forgot the cuda synchronize().

Work.wait() syncs one cuda stream used for comms back to the default cuda stream, it does Not sync to the CPU.

Reply:
@wconstab do you mean an additional call of torch.cuda.synchronize(device=device)? This did not work in my  and the tests still fail. 

Also in normal pg collectives, Python users are expected to call both wait() and cuda synchronize() if he/she want the collective to be a sync call? This seems bad design for wait()

Reply:
Yes, that's the way cuda/pytorch works to be honest.

Even if you call torch.matmul() and then exit your test, the matmul might still be running on you. You have to explicitly sync gpu if you want to make sure it's finished.

The reason we don't do a sync inside work.wait is because most often that is not what users want and would be slower. Instead it helps ensure that collective finishes before any new you kernels start. It helps manage dependencies between different streams of cuda ops.